src/storage-client/src/controller/remap_migration.rs

```diff
@@ -0,0 +1,125 @@
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use std::collections::{BTreeMap, BTreeSet};
+
+use differential_dataflow::lattice::Lattice;
+
+use timely::order::TotalOrder;
+use timely::progress::Timestamp;
+
+use mz_ore::now::EpochMillis;
+use mz_persist_types::Codec64;
+use mz_proto::RustType;
+use mz_repr::{GlobalId, TimestampManipulation};
+
+use crate::client::{StorageCommand, StorageResponse};
+use crate::controller::{ProtoStorageCommand, ProtoStorageResponse};
+use crate::types::sources::IngestionDescription;
+
+use super::{
+    CollectionDescription, Controller, DataSource, DurableCollectionMetadata, StorageController,
+};
+
+impl<T> Controller<T>
+where
+    T: Timestamp + Lattice + TotalOrder + Codec64 + From<EpochMillis> + TimestampManipulation,
+    StorageCommand<T>: RustType<ProtoStorageCommand>,
+    StorageResponse<T>: RustType<ProtoStorageResponse>,
+
+    Self: StorageController<Timestamp = T>,
+{
+    /// Determine the delta between `durable_metadata` and `collections` such
+    /// that:
+    /// - Each remap collection's data shard is its parent collection's remap
+    ///   shard.
+    /// - No collection contains a `Some` value for its remap shard.
+    ///
+    /// Apply this delta using
+    /// `Controller::upsert_collection_metadata(durable_metadata, <this
+    /// function's return value>)`.
+    ///
+    /// This approach is safe/backward compatible because:
+    /// - Every ingestion collection previously had a remap shard
+    /// - Every ingestion collection now has a progress collection subsource,
+    ///   whose data shard should be the remap shard
+    /// - No other type of collection used their remap shards, so dropping them
+    ///   entirely is essentially a nop.
+    ///
+    /// MIGRATION: v0.44
+    pub(super) fn remap_shard_migration(
+        &mut self,
+        durable_metadata: &BTreeMap<GlobalId, DurableCollectionMetadata>,
+        collections: &[(GlobalId, CollectionDescription<T>)],
+    ) -> BTreeMap<GlobalId, DurableCollectionMetadata> {
+        let mut state_to_update = BTreeMap::new();
+
+        let mut progress_collections_pending = BTreeSet::new();
+
+        for (id, desc) in collections {
+            // Track that we might be adding metadata for progress collections.
+            if matches!(desc.data_source, DataSource::Progress) {
+                progress_collections_pending.insert(*id);
+            }
+
+            let mut current_metadata = match durable_metadata.get(id) {
+                Some(c) => {
+                    // If we see any existing collection with a retired remap
+                    // shard, we know the migration has already completed.
+                    if c.remap_shard.is_none() {
+                        return state_to_update;
```

Comment:
Why do we still return `state_to_update` and not an empty `BTreeMap`? The comment seems to say that we don't have to do _any_ migration when we hit this case.

Reply:
Ah, that was an oversight. Will fix.